### ISSUE TITLE
Can prepend PATH in kit?

### ISSUE THREAD

**Issue by pbl-pw** [microsoft/vscode-cmake-tools#460]
I have two mingw : C:\mingw32 and C:\mingw64
so not good to add both to PATH.
Add below to cmake-kits.json but not work:
```
	{
		"name": "mingw64",
		"compilers": {
			"CC": "c:/mingw64/bin/gcc.exe",
			"CXX": "c:/mingw64/bin/g++.exe"
		},
		"environmentVariables": {
			"PATH":"C:\\mingw64\\bin;${env:PATH}"
		}
	}
```

**Comment by KoeMai** [comment]
yes

**Comment by vector-of-bool** [comment]
Looks like a separate `prependPath` and/or `appendPath` option to kits might be of good use...

**Comment by bjude** [comment]
I'd like this too, but as a workspace local option not just for kits.
Often i need to expose specific versions of libraries to a cmake program that will cause big headaches to simple add to the system PATH variable and currently i'm at a loss for how to do that.

Adding the below modifies the path, but seems to prevent the kits adding the required paths (the MSVC paths aren't added if i do this)
`"cmake.configureEnvironment": {
     "PATH": "${env:Path};C:/code/libFoo"
 }`

**Comment by lygstate** [comment]
Don't know if this issue are fixed

**Comment by vector-of-bool** [comment]
`prependPath` and `appendPath` may still be useful. The existing MinGW kit's `PATH` variable might be rewritten in terms of this.

**Comment by lygstate** [comment]
Leave this open.

**Comment by lygstate** [comment]
Refer to https://github.com/vector-of-bool/vscode-cmake-tools/pull/596

**Comment by pm-energid** [comment]
> I'd like this too, but as a workspace local option not just for kits.
> Often i need to expose specific versions of libraries to a cmake program that will cause big headaches to simple add to the system PATH variable and currently i'm at a loss for how to do that.
> 
> Adding the below modifies the path, but seems to prevent the kits adding the required paths (the MSVC paths aren't added if i do this)
> `"cmake.configureEnvironment": { "PATH": "${env:Path};C:/code/libFoo" }`

I also need to prepend some directories to the path when building using CMake. Encountering thee same issue that the MSVC paths are overridden if I specify the environments from cmake tools settings. Options for prepending or appending to existing variables would certainly be helpful.

**Comment by andreeis** [comment]
@pm-energid, we have a fix for the path override problem that will be included in 1.7. Also there is the possibility to run a script when setting your kit (try "environmentSetupScript" in the definition of the kit you are using).
Even before we release 1.7 you can try the latest vsix of this extension, for example from here (look at the vsix link at the bottom): https://github.com/microsoft/vscode-cmake-tools/actions/runs/687275647.
Let us know if all these unblock you.

**Comment by pm-energid** [comment]
> @pm-energid, we have a fix for the path override problem that will be included in 1.7. Also there is the possibility to run a script when setting your kit (try "environmentSetupScript" in the definition of the kit you are using).
> Even before we release 1.7 you can try the latest vsix of this extension, for example from here (look at the vsix link at the bottom): https://github.com/microsoft/vscode-cmake-tools/actions/runs/687275647.
> Let us know if all these unblock you.

Hi Andreea,

How should I try this newer version? I installed it, but don't see any option changes. Are the changes in behavior of how CMake environment, CMake build environment, CMake configure environment work?

Also, I didn't see anything about "environmentSetupScript" in the CMake Tools documentation. I do see some references to it in certain issues though.

Thanks,
Piyoosh

**Comment by andreeis** [comment]
@pm-energid, yes we should explain environmentSetupScript. Our documentation needs some updates in general.

Regarding execution of a script before any call to cmake. Does it really have to be every time before any cmake invocation (configure, build, ...etc) or once at load time is enough, as long as the environment changes are persisted throughout the entire VSCode session? environmentSetupScript is meant to keep the environment effects of running it for all the future configure/build operations during the current VSCode session. Let us know if that's not the case (it would be a bug that we'd want to fix).

I am glad that trying the latest vsix confirmed that the environment settings expansion bug was fixed. So, are you completely unblocked and your scenario works or there's something else you'd like us to address?
